### ISSUE TITLE
Strange default behavior for output shape

### ISSUE THREAD

**Issue by csinva** [pytorch/pytorch#19991]
If Y is one-dimensional, the Y vector must have shape (batch_size, 1) and not just shape (batch_size). If it does not have the 2nd dimension of size 1, the default behavior is very misleading.

A simple script reproducing this behavior is located [here](https://github.com/csinva/pytorch_dangerous_mistakes/blob/master/output_shape_err.ipynb).

**Comment by soumith** [comment]
That behavior is called broadcasting, and it's what we implemented looking at NumPy. Here's a link to read: https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html

Where it gets particularly error-prone is inputs to loss functions (as your example encounters). We now issue a warning as of release v1.1 when such a situation occurs, so that the user is aware of and can fix the issue.
